apispec/ext/marshmallow/openapi.py

```diff
@@ -603,13 +621,11 @@ class Meta:
             ):
                 continue
 
-            def prop_func(field_obj=field_obj):
-                return self.field2property(
-                    field_obj, use_refs=use_refs, dump=dump, load=load, name=name,
-                )
-
             observed_field_name = self._observed_name(field_obj, field_name)
-            jsonschema['properties'][observed_field_name] = prop_func
+            property = self.field2property(
+                field_obj, use_refs=use_refs, dump=dump, load=load, name=name,
+            )
+            jsonschema['properties'][observed_field_name] = property
```

Comment:
AFAIU, this change and the change above (`OrderedLazyDict()` -> `OrderedDict()`) are related to each other but unrelated to the issue, right?

Last time I checked, I didn't understand this part of the code, so I appreciate that you could make it simpler but I can't comment on the impact.

Reply:
No these changes are necessary if we want to add a schema to the spec from within `field2property`.  The way it worked before is that for each field a `prop_func` was created with a default parameter of the `field_obj`.  Then that `prop_func` is loaded into a `LazyDict`.  A the end of `fields2jsonschema` the `LazyDict` is returned but _none of the `prop_func`s are called_.  They only get called when accessing the item in the dict.  

This strategy was added [here](https://github.com/marshmallow-code/apispec/commit/fa1ec525fcd1d840e5d1e955ab8cfdfc11367943).  With the plan being that if execution of `field2property` was delayed until all of the schemas are in the spec (and in the ref dictionary) then when `field2property` eventually gets called it will find the the nested schemas in the ref, insert a ref into the spec, and stop infinite recursion.

Now if we want to move the functionaity to add a nested schema to the spec in `field2property` there is a problem with this approach.  The first time iterating over the spec's definitions all of the all of the `prop_func`s in the spec get called for the first time and `field2property` tires to add schemas to the dictionary during iteration which throws an error.  For some reason things like `print spec.to_dict()` will work but `json.dumps(spec.to_dict())` will not.  

I think the original authors of the autoreferencing functionality were running into this, which is why they put they logic to iterate over the fields and add nested schemas to the spec where they did.  In `definition_helper` `inspect_schema_for_autoreferencing` was adding all of the nested schemas into the spec before any of the `field2property` functions are called.  

The plan in this pr is to use regular dictionaries and call `field2property` immediately from `fields2jsonschema`.  All nested schemas will get added to the spec along the way via a call to `spec.definition` from `field2property` - instead of manually in the original plan or via `inspect_schema_for_autoreferencing`.  

I hope this clarifies things, but I fear it might not...